src/mlpack/core/data/load_utilities.hpp

```diff
@@ -0,0 +1,283 @@
+/**
+ * @file core/data/load_impl.hpp
+ * @author Ryan Curtin
+ * @author Omar Shrit
+ * @author Gopi Tatiraju
+ *
+ * Implementation of templatized load() function defined in load.hpp.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_LOAD_UTILITIES_HPP
+#define MLPACK_CORE_DATA_LOAD_UTILITIES_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <string>
+
+#include "dataset_mapper.hpp"
+#include "detect_file_type.hpp"
+#include "format.hpp"
+
+namespace mlpack {
+namespace data {
+
+namespace details{
+
+template<typename Tokenizer>
+std::vector<std::string> ToTokens(Tokenizer& lineTok)
+{
+  std::vector<std::string> tokens;
+  std::transform(std::begin(lineTok), std::end(lineTok),
+                 std::back_inserter(tokens),
+                 [&tokens](std::string const &str)
+  {
+    std::string trimmedToken(str);
+    Trim(trimmedToken);
+    return std::move(trimmedToken);
+  });
+
+  return tokens;
+}
+
+inline
+void TransposeTokens(std::vector<std::vector<std::string>> const &input,
+                     std::vector<std::string>& output,
+                     size_t index)
+{
+  output.clear();
+  for (size_t i = 0; i != input.size(); ++i)
+  {
+    output.emplace_back(input[i][index]);
+  }
+}
+} // namespace details
+
+/**
+ * All possible DataOptions grouped under one class.
+ * This will allow us to have consistent API over the next mlpack versions. If
+ * new data options might be necessary, then they should be added in the
+ * following.
+ */
+class LoadOptions
+{
+ public:
+  /**
+   * DataOptions constructor that takes the followint parameters.
+   *  - fatal : true if fatal error should be reported.
+   *  - hasHeaders : true if the dataset file has headers we need to recover
+   *  - transpose: true by default, Transpose armadillo matrix to column major
+   *  - semiColon: If the dataset separator is a semicolon instead of commas.
+   *  - missingToNan: replace missing values to NaN.
+   *  - categorical: if the dataset contains categorical values.
+   *  - image: true if we are trying to load an image.
+   *  - fileFormat: Detect automaticaly the file type, csv, txt, hdf5.
+   *  - objectName: Model / object name that is going to be serialized
+   *  - imgInfo: Information about the image if we already have them.
+   *  - dataFormat: the data serialization format: xml, bin, json
+   */
+  LoadOptions(
+      bool fatal = true,
+      bool hasHeaders = false,
+      bool transpose = true,
+      bool semiColon = false,
+      bool missingToNan = false,
+      bool categorical = false,
+      bool image = false,
+      bool model = false,
+      FileType fileFormat = FileType::AutoDetect,
+      format dataFormat = format::binary,
```

Comment:
One of the "scary" things about putting the serialization formats and also the matrix file loading formats into one struct is that a user could write code like this:

```
data::Load("my_file.csv", matrix, LoadOptions::Format::CSV | LoadOptions::DataFormat::Binary)
```

So I wonder if it might be nicer to split into two different `LoadOptions` classes, something like `ModelLoadOptions` that only has `format` and `fatal`.  This of course is not the only possible way to do it, but my "fear" is that when we use only one struct to represent all the options, then it is not possible to throw a *compiler* error when users specify options that don't go with the format/file type/object type that they're loading.  In either case, we do need to have some form of error handling for when users specify options that don't make sense or conflict or something like that.

Also, I can't put this comment any lower, but to determine whether or not 

```
// Uses SFINAE to detect when we are loading a model instead of a matrix.
template<typename T>
bool Load(const std::string& filename,
                   T& object,
                   LoadOptions& opts,
                   // you would need the opposite of this condition in the matrix load function
                   const std::enable_if<!arma::is_arma_type<T>::value && HasSerialize<T>::value>::type* = 0);
```